lib/active_fedora/indexing/descendant_fetcher.rb

```diff
@@ -0,0 +1,104 @@
+    # with certain hasModel models. This feature is used in some hydra apps that need to
+    # index 'permissions' objects before other objects to have the solr indexing work right.
+    # And so by default, the prioritized class names are the ones form Hydra::AccessControls,
+    # but you can alter the prioritized model name list, or set it to the empty array.
+    #
+    #     DescendentFetcher.new(ActiveFedora.fedora.base_uri).descendent_and_self_uris
+    #     #=> array including self uri and descendent uris with "prioritized" (by default)
+    #         Hydra::AccessControls permissions) objects FIRST.
+    #
+    # Change the default prioritized hasModel names:
+    #
+    #     ActiveFedora::Indexing::DescendantFetcher.default_priority_models = []
+    class DescendantFetcher
+      HAS_MODEL_PREDICATE = ::RDF::URI.new('info:fedora/fedora-system:def/model#hasModel').freeze
+
+      define_singleton_method(:default_priority_models) do
+        @default_priority_models ||= %w(Hydra::AccessControls Hydra::AccessControls::Permissions).freeze
+      end
+      define_singleton_method(:default_priority_models=) do |arg|
+        @default_priority_models = arg
+      end
+
+      attr_reader :uri, :priority_models
+
+      def initialize(uri,
+                     priority_models: self.class.default_priority_models)
+        @uri = uri
+        @priority_models = priority_models
+      end
+
+      def descendant_and_self_uris
+        partitioned = descendant_and_self_uris_partitioned
+        partitioned[:priority] + partitioned[:other]
+      end
+
+      # returns a hash where key :priority is an array of all prioritized
+      # type objects, key :other is an array of the rest.
+      def descendant_and_self_uris_partitioned
+        resource = Ldp::Resource::RdfSource.new(ActiveFedora.fedora.connection, uri)
+        # GET could be slow if it's a big resource, we're using HEAD to avoid this problem,
+        # but this causes more requests to Fedora.
+        return partitioned_uris unless resource.head.rdf_source?
+
+        add_self_to_partitioned_uris
+
+        immediate_descendant_uris = rdf_graph.query(predicate: ::RDF::Vocab::LDP.contains).map { |descendant| descendant.object.to_s }
+        immediate_descendant_uris.each do |descendant_uri|
+          self.class.new(
+            descendant_uri,
+            priority_models: priority_models
+          ).descendant_and_self_uris_partitioned.tap do |descendant_partitioned|
+            partitioned_uris[:priority].concat descendant_partitioned[:priority]
+            partitioned_uris[:other].concat descendant_partitioned[:other]
+          end
+        end
+        partitioned_uris
+      end
+
+      protected
+
+        def rdf_resource
+          @rdf_resource ||= Ldp::Resource::RdfSource.new(ActiveFedora.fedora.connection, uri)
+        end
+
+        def rdf_graph
+          @rdf_graph ||= rdf_resource.graph
```

Comment:
I don't know of any way to do this prioritization without pulling it. Unless your suggestion above of `ActiveFedora::Base.find(ActiveFedora::Base.uri_to_id(uri))` will let me just look at the ruby class of the pulled object, instead of the object graph? Is that true, will `ActiveFedora::Base.find` return a ruby object classed `Hydra::AccessControls::Permission` in the relevant circumstance?

Otherwise, we could avoid doing the check at all if the `priority_models` is set to nil or the empty array, so at least the client could turn off the expensive operation if they have no need of prioritization.